### PR TITLE
Added unit tests for the Metadata class

### PR DESCRIPTION
**Pull request by robertboloc**
This pull requests adds unit tests, with 100% coverage, for the Metadata class.

### CODE REVIEW

File: src/Metadata/Metadata.php [hunk 1]
```diff
@@ -41,6 +41,7 @@ public function __construct(Adapter $adapter)
      *
      * @param  Adapter $adapter
      * @return Source\AbstractSource
+     * @throws \Exception
```
Comment (Maks3w) [inline src/Metadata/Metadata.php]: Add description when this exception is thrown

File: test/Metadata/MetadataTest.php [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+        );
+    }
+
+    /**
+     * @dataProvider createSourceFromAdapterDataProvider
+     *
+     * @param $platform
+     * @param $sourceExpectation
+     */
+    public function testCreateSourceFromAdapterCreatesPlatformSpecificSource($platform, $sourceExpectation)
```
Comment (Maks3w) [inline test/Metadata/MetadataTest.php]: This test is redundant with `testTheConstructorSetsTheAdaptersAndTheSource`

First one could use the data provider

File: test/Metadata/MetadataTest.php [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+        $adapterReflectionProperty = new \ReflectionProperty($metadata, 'adapter');
+        $adapterReflectionProperty->setAccessible(true);
+        $adapterPropertyValue = $adapterReflectionProperty->getValue($metadata);
+
+        $sourceReflectionProperty = new \ReflectionProperty($metadata, 'source');
+        $sourceReflectionProperty->setAccessible(true);
+        $sourcePropertyValue = $sourceReflectionProperty->getValue($metadata);
+
+        $this->assertSame($adapterMock, $adapterPropertyValue);
```
Comment (Maks3w) [inline test/Metadata/MetadataTest.php]: Place this assertion after adapterProperyValue assignment

Comment (Maks3w) [inline test/Metadata/MetadataTest.php]: Better if use `$this->assertAttributeSame()` and remove all that reflection logic

File: test/Metadata/MetadataTest.php [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+        $this->assertInstanceOf(
+            'Zend\Db\Metadata\Source\MysqlMetadata',
+            $sourcePropertyValue
+        );
+    }
+
+    /**
+     * @dataProvider createSourceFromAdapterDataProvider
+     *
+     * @param $platform
```
Comment (Maks3w) [inline test/Metadata/MetadataTest.php]: add param types or remove the tag

**Comment by robertboloc** [comment]
@Maks3w I have update the pull request with your feedback from the code review, thanks!

File: test/Metadata/MetadataTest.php [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+     * @param string           $sourceExpectation
+     */
+    public function testTheConstructorSetsTheAdaptersAndTheSource($platform, $sourceExpectation)
+    {
+        /** @var \Zend\Db\Adapter\Adapter $adapterMock */
+        $adapterMock = $this->getMockBuilder('Zend\Db\Adapter\Adapter')
+                            ->disableOriginalConstructor()
+                            ->getMock();
+
+        $adapterMock->expects($this->once())
```
Comment (Maks3w) [inline test/Metadata/MetadataTest.php]: Could you extract all the thing about `adapterMock` in a method called `createAdapterMockWithPlatform($platform)`

I think will improve test readiblity.

File: test/Metadata/MetadataTest.php [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+            '',
+            false,
+            true,
+            true,
+            [$proxyMethod]
+        );
+
+        $sourceMock->expects($this->once())
+                   ->method($proxyMethod)
+                   ->willReturn($returnValue);
```
Comment (Maks3w) [inline test/Metadata/MetadataTest.php]: Could you test arguments received too?

**Comment by Maks3w** [comment]
I start to think this class should be deprecated and only to have a factory class with a method for get the source from adapter

**Comment by robertboloc** [comment]
I am not actually using the class. I just saw it had no coverage and I had some spare time so I decided to write tests for it. It does look like the only useful bit is the switch to get the source instance, as all the rest of the methods just proxy to the source.

I will leave the pull request as is for now, until a decision is made

**Comment by Maks3w** [comment]
All the thing about to test proxy methods reveal a bad design flow.

**Comment by mwillbanks** [comment]
@Maks3w that seems fair; however, i don't think it a bad idea to add the unit tests that have been add here.  Actually it is very well likely a great idea for the time being.  I would say we should merge the PR and then we can deprecate the other part in the future.

**Comment by Maks3w** [comment]
I think I wrote a PoC with a more clean structure let me to dig on it

**Pull request closed without merging**